Vary whether to left or right align PGM allocations.

nishajain61 opened the pull request:
#### a2ce7711e418202bd84ca34e529cad3f9f05850e
<pre>
Vary whether to left or right align PGM allocations.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266581">https://bugs.webkit.org/show_bug.cgi?id=266581</a>
<a href="https://rdar.apple.com/107954062">rdar://107954062</a>.

Reviewed by David Kilzer.

Large heap PGM allocations are right aligned to catch overflows by default.
Added the left alignment with random decision in order to catch underflows with equal probability.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate): Added random decision for left or right alignment.
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h: Added flag in meta data to keep track of alignment for given heap.
/Source/bmalloc/libpas/src/libpas/pas_report_crash.c: Modified the crash reporting based on alignment.
/Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h: Added alignment info.

Canonical link: <a href="https://commits.webkit.org/279892@main">https://commits.webkit.org/279892@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81498458e045f357c32d91d322344b336b3ce1e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54809 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34245 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7388 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58087 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5540 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41790 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5556 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44390 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3747 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56904 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32370 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47483 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25514 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29159 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4819 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3681 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/48176 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50988 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5037 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59677 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/54314 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30057 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5185 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51813 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31198 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47564 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51230 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32208 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66610 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8122 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30986 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12692 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -97,57 +102,103 @@ kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, m
+            addr64_t bottom = (addr64_t) (key - pgm_metadata->mem_to_waste - pgm_metadata->page_size);
+            addr64_t top = (addr64_t) (key - pgm_metadata->mem_to_waste);
+
+            if (pas_fault_address_is_in_bounds(fault_address, bottom, top)) {
```
I still feel we have a lot of code duplication here:

Lines 113-117 are duplicated at 140-144, 158-161, and 182-185.
Lines 125-128 are duplicated at 170-173.

We can create functions or macros to help limit the amount of code duplication.

stwrt reviewed (commented):
Thank you for abstracting out the crash field setting.

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
Nit:  This comment isn't needed.

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
I think you are missing a `return` statement in every call to `pas_update_report_crash_fields()` (only noting this once):

```suggestion
            if (pas_fault_address_is_in_bounds(fault_address, bottom, top))
                return pas_update_report_crash_fields(report, "out-of-bounds", "high", fault_address, pgm_metadata->allocation_size_requested);
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -53,6 +53,19 @@ static memory_reader_t * setup_memory_reader(crash_reporter_memory_reader_t crm_
+    return (fault_address >= bottom) && (fault_address < top);
+}
+
+static PAS_ALWAYS_INLINE kern_return_t pas_update_report_crash_fields(pas_report_crash_pgm_report *report, const char *error_type, const char *confidence, vm_address_t fault_address, size_t allocation_size)
```
This function should also have `__attribute__((__warn_unused_result__))` based on the way it's used below.

```suggestion
static PAS_ALWAYS_INLINE kern_return_t __attribute__((__warn_unused_result__)) pas_update_report_crash_fields(pas_report_crash_pgm_report *report, const char *error_type, const char *confidence, vm_address_t fault_address, size_t allocation_size)
```

—

Might also want to define `PAS_WARN_UNUSED_RETURN` in `Source/bmalloc/libpas/src/libpas/pas_utils.h`:

```
#define PAS_WARN_UNUSED_RETURN __attribute__((__warn_unused_result__))
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
This comment should be:

```suggestion
            // Right-aligned "Upper PGM OOB" checking
            // FIXME: This assumes the allocation is still live.
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
Add FIXME comment about liveness check:

```suggestion
            // Right-aligned "Lower PGM OOB" checking
            // FIXME: This assumes the allocation is still live.
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
This comment should be:

```suggestion
            // Right-aligned "UAF + OOB" checking towards lower guard page
            // FIXME: This assumes the allocation is not live.
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
This is a low-confidence long-range OOB:

```suggestion
                pas_update_report_crash_fields(report, "long-range OOB", "low", fault_address, pgm_metadata->allocation_size_requested);
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
Let's make this comment less verbose:

```suggestion
            // Left-aligned "Lower PGM OOB" checking
            // FIXME: This assumes the allocation is still live.
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
Update comment:

```suggestion
            // Left-aligned "UAF + OOB" checking towards upper guard page
            // FIXME: This assumes the allocation is not live.
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
This should check the "wasted memory" region:

```suggestion
            bottom = (addr64_t) (key + pgm_metadata->size_of_data_pages - pgm_metadata->mem_to_waste);
            top = (addr64_t) (key + pgm_metadata->size_of_data_pages);
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
This should check the "wasted memory" region:

```suggestion
            bottom = (addr64_t) (key - pgm_metadata->mem_to_waste);
            top = (addr64_t) key;
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```suggestion
            // Left-aligned "Upper PGM OOB" checking
            // FIXME: This assumes the allocation is still live.
```

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
This is a low-confidence `long-range OOB`:

```suggestion
                pas_update_report_crash_fields(report, "long-range OOB", "low", fault_address, pgm_metadata->allocation_size_requested);
```

ddkilzer reviewed (approved):
Overall this is great!  I like the way the logic is broken out between right and left alignment now—it's much easier to read.

There are a lot of requested changes, but if you make them all, this can land.  (I'll do one more check before I add the commit-queue label.)

Overview of requested changes:
- Add attribute to warn when the result of `pas_fault_address_is_in_bounds()` is not used to catch errors in the future.
- Update comments to use consistent terminology and add FIXMEs for allocation liveness.
- Add `return` statements where `pas_fault_address_is_in_bounds()` is used.
- Introduce the "long-range OOB" concept and adjust the bounds check for it.
- "OOB + UAF" checks should only check the "wasted memory" region (since the "long-range OOB" covers the far guard page).

nishajain61 committed:
Vary whether to left or right align PGM allocations.
https://bugs.webkit.org/show_bug.cgi?id=266581
rdar://107954062.

Reviewed by David Kilzer.

Large heap PGM allocations are right aligned to catch overflows by default.
Added the left alignment with random decision in order to catch underflows with equal probability.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate): Added random decision for left or right alignment.
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h: Added flag in meta data to keep track of alignment for given heap.
/Source/bmalloc/libpas/src/libpas/pas_report_crash.c: Modified the crash reporting based on alignment.
/Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h: Added alignment info.

Canonical link: https://commits.webkit.org/279892@main
